### ISSUE TITLE
[Feature Request] Dynamically adjusting thread pool as ram availability increases/decreases

### ISSUE THREAD

**Issue by Caffiendish** [fastbuild/fastbuild#959]
A useful feature for FastBuild to have, would be dynamically adjusting the threadpool, so that, if we're at less than 500MB of ram free, for example, we don't spawn more compiler threads, and lock up the system entirely.
Or conversely, if more ram becomes available, we could launch more threads, and blast through some extra tasks.

I was trying to compile Unreal from source with FBuild present leads to my local system completely locking up, and Windows failing entirely, because FBuild doesn't take into account the requested ram per action (probably because it doesn't know/get told).

Here's a log excerpt from VS, where (I assume) UnrealBuildTool discerns that terrible choices are about to be made, and course corrects:
> 1>Determining max actions to execute in parallel (16 physical cores, 32 logical cores)
1>  Executing up to 16 processes, one per physical core
1>  Requested 1.5 GB free memory per action, 16.82 GB available: limiting max parallel actions to 11

Strangely, UBT will respect processor limitation requests if FBuild isn't present, but since it doesn't pass along the -j param to FBuild, that's something to work out with UBT.

If I actually could use FBuild's distributed building for Unreal, having all threads firing would be really useful for shader compilation, but locking up every machine because I forgot that I'm about to build an absolute ram demon (or the correct variables aren't being sent to FBuild) is slightly less helpful, and it'd be incredibly useful if Fastbuild would take mercy upon us plebs in these cases.

**Comment by ffulin** [comment]
I don't believe a dynamic throttle would be very effective because the peak memory usage of a particular process is not known up-front and varies significantly from invocation to invocation and during the lifetime of the process. At the point where we would decide to spawn another process or not the total memory usage would often be very low. After the processes run for a while the memory use would become too high, and a that point it's too late. The start of the build being the most obvious problematic case, but this problem would be an issue throughout builds.

I think explicitly reducing the concurrent counts via -j in situations where you know the combination of code base and hardware is mismatched is the right way to tackle this problem.

I think as you've noted in the Unreal case, since the UBT is used as a wrapper/layer above FASTBuild and already has logic driven around its knowledge of the specifics of the Unreal codebase, it makes sense for the UBT to pass the appropriate -j parameter to match the throttling it would pass to other build systems already.

I don't think it makes sense for FASTBuild to try and emulate that behavior itself either, since it works with a variety of codebase, most of which are much "leaner" than Unreal.

**> but locking up every machine because I forgot that I'm about to build an absolute ram demon**

Workers have options available (in the UI or via command line) to limit their use of machine resources and can be tuned to operate assuming ownership of the machine (Dedicated mode) or share use (Idle mode). They also have some additional disk space and low memory condition detection to reduce resource exhaustion likelyhood (though imperfect for the same reasons as mentioned above)

There may be opportunities to further improve the logic of workers, particularly when operating in "when idle" mode.

**Comment by Caffiendish** [comment]
I hadn't seen the command line options for the worker (was very focused on fbuild), that's very helpful, and I'll bear that in mind when using my main machine for builds, thanks!

I've submit a PR for fixing the UBT, which is currently happily building UE on this machine with Fastbuild spawning a reasonable amount of threads. 👍 

UBT was clearly at fault here, but IMHO this highlights an issue with the local worker/manager, given that it has no logic for zero ram availability; FBuild was continuing to launch _multiple_ compiler processes when at _sustained_ 100% committed ram.

I can totally understand launching a process _under_ 100% usage, it's not Fastbuild's place to know how much ram might be needed (that'd just be convenient functionality), but I feel like it _should_ be able to discern that launching multiple compiler threads when there's no resources available _could_ be a blunder.

**Comment by ffulin** [comment]
**> I've submit a PR for fixing the UBT,** 
Nice one :)

**> I feel like it should be able to discern that launching multiple compiler threads when there's no resources available could be a blunder.**

It would be possible to add this logic, but it would make all task spawning more expensive with the cost of this logic exceeding the cost of some tasks. It would also complicate the task scheduler which is currently event driven and free of busy waits (this logic would require some sort of sleep/delay logic). With more thought or perhaps additional complexity, some of the performance costs might be avoided or minimized.

The question in my mind is whether those costs are worth the benefits. If this logic existed right now it would not tangibly improve the situation you experienced and your machine would still be  unusable or unstable as tasks that eventually exceed ram limits would frequently be spawned throughout your compile.

I suppose it can be argued that it would be less bad overall, but when this situation only has to happen once in a build to be somewhat catastrophic and this would prevent only a subset of occurrences for a given builds, I still feel it's perhaps not a worthwhile tradeoff.

It could also be argued that logic like this would hide what is a severe performance problem hidden by under spec'd hardware. I think personally I prefer that I can see physical ram is being exhausted and decide on what action to take (limit core count or upgrade ram) as opposed to just having slower builds with no obvious cause. Perhaps this downside could be mitigated by improvements to build profiling functionality (like -profile)

I'm not completely closed off to this idea, but I think on balance given the downsides and how I think it would currently need to be implemented I think I wouldn't want to add this throttling. I'm going to continue thinking about this though - perhaps some new ideas might alter than tradeoffs.

**Comment by flyingllama87** [comment]
Hello,

<3 the fbuild project so please don't think I'm ungrateful.

We've just ran into this issue (Unreal Engine too) and will address it by using the -j option. That said, it'd be great if fbuild was a bit more clever in this regard and at least notify users about the potential RAM exhaustion. Most of the devs running into this issue where I work responded to this by just turning off fastbuild as opposed to looking into why fastbuild was slowing down their builds/PCs.

I think this issue is exacerbated by intel 12th gen+ CPUs that have both performance and efficiency cores, so the systems have a high number of available threads but the systems may not have the RAM to match. An alternative fix is to upgrade everyone's RAM, but that may not be viable to all users.

Thanks!

**Comment by Caffiendish** [comment]
If there's other people that happen to stumble across this, the fix was accepted into the Unreal codebase, but it's only just made it into the 5.3 branch:

https://github.com/EpicGames/UnrealEngine/blob/5.3/Engine/Source/Programs/UnrealBuildTool/Executors/Experimental/FASTBuild.cs#L1340

So that should be included in the binary builds going forwards.
For previous builds downloaded from the launcher, you could just build the UBT with that line and then replace the binary, I don't think you need to rebuild the _entire_ engine.

**Comment by karofsky** [comment]
> If there's other people that happen to stumble across this, the fix was accepted into the Unreal codebase, but it's only just made it into the 5.3 branch:
> 
> https://github.com/EpicGames/UnrealEngine/blob/5.3/Engine/Source/Programs/UnrealBuildTool/Executors/Experimental/FASTBuild.cs#L1340
> 
> So that should be included in the binary builds going forwards. For previous builds downloaded from the launcher, you could just build the UBT with that line (and add the variable to the line on 1347) and then replace the binary, I don't think you need to rebuild the _entire_ engine.

this works.

**Comment by ffulin** [comment]
On a related note, the next version of FASTBuild (v1.14) contains a new "[Concurrency Groups](https://github.com/fastbuild/fastbuild/commit/0d4ba7abd086bc17f4cf5d4664a03057647573d6)" feature which allows for limits to be applied to certain jobs based that are more strict than the overall pool size (limiting concurrency of certain tasks either absolutely or by system memory).

This might be possible for the UBT to leverage to have better control/performance.

**Comment by tru** [comment]
Hello,

Thanks for the addition of Concurrency Groups! We ran into something very similar to this. In our case we have some really big and heavy LTO linker jobs. We used this feature to put all the LTO jobs into a single concurrency group and then limit it to 1. But this doesn't work if you have a larger number of compile jobs and linker jobs. What ends up happening for us is that the LTO job is scheduled together with a bunch of compilers. This leads to OOM on our machine since the LTO job is using a lot of memory and CPU (it's multithreaded).

We are thinking about solutions here - we would like a way to tell the fastbuild scheduler that a command is multithreaded so that it holds back on scheduling them all at once. I.e. you set `.MultiThreaded = 100%` or something similar - then the scheduler would limit these jobs until there is "space" for it and limit scheduling new stuff. Not to familiar with the jobscheduler yet but I will investigate that.

Another option is to somehow make it so that compile jobs finish first before *any* linker jobs are scheduled. Then we can control it with the concurrency groups. Maybe so that the concurrency groups are dependent on each other? But that seems a bit messy.

Do you have any thoughts on this @ffulin ?
